proposals/0177-program-runtime-abiv2.md

```diff
@@ -0,0 +1,156 @@
+---
+simd: '0177'
+title: Program Runtime ABI v2
+authors:
+  - Alexander Meißner
+category: Standard
+type: Core
+status: Idea
+created: 2025-02-23
+feature: TBD
+extends: SIMD-0219
+---
+
+## Summary
+
+Align the layout of the virtual address space to large pages in order to
+simplify the address translation logic and allow for easy direct mapping.
+
+## Motivation
+
+At the moment all validator implementations have to copy (and compare) data in
+and out of the virtual memory of the virtual machine. There are four possible
+account data copy paths:
+
+- Serialization: Copy from program runtime (host) to virtual machine (guest)
+- CPI call edge: Copy from virtual machine (guest) to program runtime (host)
+- CPI return edge: Copy from program runtime (host) to virtual machine (guest)
+- Deserialization: Copy from virtual machine (guest) to program runtime (host)
+
+To avoid this a feature named "direct mapping" was designed which uses the
+address translation logic of the virtual machine to emulate the serialization
+and deserialization without actually performing copies.
+
+Implementing direct mapping in the current ABI v0 and v1 is very complex
+because of unaligned virtual memory regions and memory accesses overlapping
+multiple virtual memory regions. Instead the layout of the virtual address
+space should be adjusted so that all virtual memory regions are aligned to
+4 GiB.
+
+## Alternatives Considered
+
+None.
+
+## New Terminology
+
+None.
+
+## Detailed Design
+
+Programs signal their support through their SBPF version field being v4 or
+above while the program runtime signals which ABI is chosen through the
+serialized magic field.
+
+### Per Transaction Serialization
+
+At the beginning of a transaction the program runtime must prepare the
+following which is shared by all instructions running programs suporting the
+new ABI. This memory region starts at `0x400000000` and is readonly. It must be
```

Comment:
Is this the memory region for the accounts? It is not clear by this sentence what is the purpose of this memory region.

I'm assuming it is, since the region for the scratchpad is the 0x5.

Reply:
> At the beginning of a transaction the program runtime must prepare the
following

That means the stuff below which is the scratchpad and the transaction accounts.

Reply:
The paragraph is difficult to understand, because you say `This memory region` without previously mentioning one. That was the source of my confusion. What do you think of this other version?

> At the beginning of a transaction the program runtime must prepare a read-only memory region starting at `0x400000000`, which is shared by all instructions running programs that support the new ABI. Throughout the transaction, it must be updated as instructions modify the account metadata or the scratchpad via `sol_set_return_data`. This memory region must contain: